### PR TITLE
Fix BSD cpu count

### PR DESCRIPTION
**Pull request by wmoxam**
`System.cpu_count` always returns -1 on BSD since fetching the number of CPUs via `sysconf` is not supported. 

Instead use `sysctl` to fetch cpu count.

### CODE REVIEW

**Comment by bew** [comment]
Maybe this should leave under `Crystal::System` as it's a function with plateform dependent implementation?

**Comment by wmoxam** [comment]
@bew There is precedence for conditional platform  dependant code elsewhere, ex: `src/process/executable_path.cr`

**Comment by bcardiff** [comment]
@wmoxam from CI there are two files where the formatter is not happy.

Error: formatting 'src/lib_c/amd64-unknown-openbsd/c/sysctl.cr' produced changes
Error: formatting 'src/lib_c/x86_64-portbld-freebsd/c/sysctl.cr' produced changes

To clarify @bew comment, #4450 is the first PR trying to move all the platform specifics in it's own namespace and offer a better abstraction. If you want to can take the stab here in this same PR. But if not a refactor will follow probably.

**Comment by wmoxam** [comment]
Okay good to know. Thanks @bcardiff

**Comment by wmoxam** [comment]
I'll refactor to match #4450

**Comment by bcardiff** [comment]
@wmoxam that PR is now merged. You can base the refactor with that final structure.
Having if flag to select which file to include and also use a skip_file macro in each file.

**Comment by wmoxam** [comment]
Thanks @bcardiff 

I'll have ready later this week

File: src/crystal/system/system.cr [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# :nodoc
```
Comment (RX14) [inline src/crystal/system/system.cr]: We don't want to `:nodoc:` `Crystal` as we loose documenting macros and `Crystal::VERSION`. The `:nodoc:` should be placed on `Crystal::System` only.

File: src/crystal/system/system.cr [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+module Crystal
+  # :nodoc
+  module System
+    # :nodoc
+    module System
```
Comment (RX14) [inline src/crystal/system/system.cr]: I think that the methods in this module could go in just `Crysatal::System` itself.

Comment (wmoxam) [inline src/crystal/system/system.cr]: Like replace `src/system.cr` with this file?

Comment (RX14) [inline src/crystal/system/system.cr]: no but place the methods in `Crystal::System` not `Crystal::System::System`. It seems redundant and these are "misc" methods which can just go in `Crystal::System` itself.

Comment (wmoxam) [inline src/crystal/system/system.cr]: IMO these are not "misc" methods, they are "System" (aka local computer system) utility methods for querying a local system's properties.  

My understanding of the structure is that `Crystal::System` is simply a wrapper namespace for platform specific implementations.

`Crystal::System::System` is meant to contain platform specific implementations of the "System" methods.

It's confusing for sure since we're using the word 'System" for two different things. Perhaps a there is a different name that is more suitable?

Comment (bcardiff) [inline src/crystal/system/system.cr]: What about `Crystal::System::Inspect`?

Comment (RX14) [inline src/crystal/system/system.cr]: @bcardiff If we changed that name we'd have to change the name of `System` to `Inspect` which would make little sense. I'd rather have bit of duplication than a nonsensical name.

**Comment by wmoxam** [comment]
@RX14 @bcardiff  How does this look now?

File: src/crystal/system/unix/hostname.cr [hunk 3]
```diff
@@ -0,0 +1,13 @@
+require "c/unistd"
+
+module Crystal::System
+  def self.hostname
+    String.new(255) do |buffer|
+      unless LibC.gethostname(buffer, LibC::SizeT.new(255)) == 0
+        raise Errno.new("Could not get hostname")
+      end
+      len = LibC.strlen(buffer)
+      {len, len}
```
Comment (RX14) [inline src/crystal/system/unix/hostname.cr]: Are you sure this string doesn't contain utf8?

File: src/crystal/system/unix/sysctl_cpucount.cr [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+module Crystal::System
+  def self.cpu_count
+    mib = Int32[LibC::CTL_HW, LibC::HW_NCPU]
+    ncpus = 0
+    size = sizeof(Int32).to_u64
+
+    if LibC.sysctl(mib, 2, pointerof(ncpus), pointerof(size), nil, 0) == 0
+      ncpus
+    else
+      -1
```
Comment (RX14) [inline src/crystal/system/unix/sysctl_cpucount.cr]: Don't we want to raise here?

File: src/crystal/system/unix/sysconf_cpucount.cr [hunk 5]
```diff
@@ -0,0 +1,9 @@
+{% skip_file if flag?(:openbsd) || flag?(:freebsd) %}
+
+require "c/unistd"
+
+module Crystal::System
+  def self.cpu_count
+    LibC.sysconf(LibC::SC_NPROCESSORS_ONLN)
```
Comment (RX14) [inline src/crystal/system/unix/sysconf_cpucount.cr]: Similarly, we should check the return value and raise.

File: src/crystal/system/unix/hostname.cr (hunk 3 above)
Comment (wmoxam) [inline src/crystal/system/unix/hostname.cr]: I didn't implement this, I only moved it 😅

File: src/crystal/system/unix/sysctl_cpucount.cr (hunk 4 above)
Comment (wmoxam) [inline src/crystal/system/unix/sysctl_cpucount.cr]: Possibly? I was just matching the behaviour of the existing `sysconf` version 

Is that the behaviour of other standard Crystal libs, or do they just rely on return value?

File: src/crystal/system.cr [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+  end
+end
+
+require "./system/unix/hostname"
+
+{% if flag?(:freebsd) || flag?(:openbsd) %}
+  require "./system/unix/sysctl_cpucount"
+{% else %}
+	# TODO: restrict on flag?(:unix) after crystal > 0.22.0 is released
+    require "./system/unix/sysconf_cpucount"
```
Comment (wmoxam) [inline src/crystal/system.cr]: Weird that the formatter didn't pick up on this

Comment (bew) [inline src/crystal/system.cr]: @wmoxam it's because the formatter doesn't format macro code

**Comment by ysbaddaden** [comment]
If changes to the API are required it may later in another pull request, but this shouldn't delay this patch which fixes issues on *BSD.

File: src/crystal/system/unix/sysconf_cpucount.cr (hunk 5 above)
Comment (wmoxam) [inline src/crystal/system/unix/sysconf_cpucount.cr]: I think this could be part of a separate PR as the return value isn't checked elsewhere. Ex: `Process.times`

**Comment by RX14** [comment]
I forgot that this was just a straight copy of the existing behaviour... Returning `-1` would be something to change in the future though.

**Comment by Sija** [comment]
[Build #4466](https://travis-ci.org/crystal-lang/crystal/builds/242443993) errored on osx with _xcode8.2_, again... Would it be a good time to merge #4528?
